### PR TITLE
incr.comp.: Let the compiler ignore incompatible incr. comp. cache artifacts

### PR DESCRIPTION
**Pull request by michaelwoerister**
Implements #35720.

cc @nikomatsakis

### CODE REVIEW

**Comment by jonathandturner** [comment]
I'm seeing travis fail with this:

```
error: environment variable `CFG_VERSION` not defined
  --> src/librustc_incremental/persist/file_format.rs:38:37
   |
38 | const RUSTC_VERSION: &'static str = env!("CFG_VERSION");
   |                                     ^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

Build failed, waiting for other jobs to finish...
error: Could not compile `rustc_incremental`.
```

**Comment by eddyb** [comment]
Blargh, I should've seen that coming. @bors r-

File: src/librustc_incremental/persist/file_format.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+/// The first few bytes of files generated by incremental compilation
+const FILE_MAGIC: &'static [u8] = b"RSIC";
+
+/// Change this if the header format changes
+const HEADER_FORMAT_VERSION: u16 = 0;
+
+/// A version string that hopefully is always different for compiler versions
+/// with different encodings of incremental compilation artifacts. Contains
+/// the git commit hash.
+const RUSTC_VERSION: &'static str = env!("CFG_VERSION");
```
Comment (michaelwoerister) [inline src/librustc_incremental/persist/file_format.rs]: Ah yes, I ran across this and then forgot about again. Should be fixed now (but let's wait for travis).

**Pull request merged**